parlai/core/image_featurizers.py

```diff
@@ -25,57 +25,98 @@ def __init__(self, opt):
         self.netCNN = None
         self.im = opt.get('image_mode', 'none')
         if self.im not in ['none', 'raw', 'ascii']:
-            self.init_cnn(self.opt)
-
-    def init_cnn(self, opt):
-        """
-        Lazy initialization of preprocessor model.
+            if 'image_mode' not in opt or 'image_size' not in opt:
+                raise RuntimeError(
+                    'Need to add image arguments to opt. See '
+                    'parlai.core.params.ParlaiParser.add_image_args'
+                )
+            self.image_mode = opt['image_mode']
+            self.image_size = opt['image_size']
+            self.crop_size = opt['image_cropsize']
+            self._lazy_import_torch()
+            self._init_transform()
+            if 'resnet' in self.image_mode:
+                self._init_resnet_cnn()
+            elif 'resnext' in self.image_mode:
+                self._init_resnext_cnn()
+            else:
+                raise RuntimeError(
+                    'Image mode {} not supported'.format(self.image_mode)
+                )
 
-        In case we don't need any image preprocessing.
-        """
+    def _lazy_import_torch(self):
         try:
             import torch
-
-            self.use_cuda = not opt.get('no_cuda', False) and torch.cuda.is_available()
-            self.torch = torch
         except ImportError:
             raise ImportError('Need to install Pytorch: go to pytorch.org')
         import torchvision
         import torchvision.transforms as transforms
         import torch.nn as nn
 
-        if 'image_mode' not in opt or 'image_size' not in opt:
-            raise RuntimeError(
-                'Need to add image arguments to opt. See '
-                'parlai.core.params.ParlaiParser.add_image_args'
-            )
-        self.image_mode = opt['image_mode']
-        self.image_size = opt['image_size']
-        self.crop_size = opt['image_cropsize']
-
+        self.use_cuda = not self.opt.get('no_cuda', False) and torch.cuda.is_available()
         if self.use_cuda:
             print('[ Using CUDA ]')
-            torch.cuda.set_device(opt.get('gpu', -1))
-
-        cnn_type, layer_num = self._image_mode_switcher()
-
-        # initialize the pretrained CNN using pytorch.
-        CNN = getattr(torchvision.models, cnn_type)
-
-        # cut off the additional layer.
-        self.netCNN = nn.Sequential(*list(CNN(pretrained=True).children())[:layer_num])
+            torch.cuda.set_device(self.opt.get('gpu', -1))
+        self.torch = torch
+        self.torchvision = torchvision
+        self.transforms = transforms
```

Comment:
interesting pattern

Reply:
this script is written as platform agnostic - i.e., it does not require that you have torch installed. So we do not import torch until someone specifies actual features to extract